lib/trino-plugin-toolkit/src/test/java/io/trino/plugin/base/cache/TestEvictableCache.java

```diff
@@ -0,0 +1,201 @@
+        }
+    }
+
+    /**
+     * Covers https://github.com/google/guava/issues/1881
+     */
+    @Test(invocationCount = 10, timeOut = TEST_TIMEOUT_MILLIS, dataProvider = "invalidations")
+    public void testInvalidateAndLoadConcurrently(Invalidation invalidation)
+            throws Exception
+    {
+        int[] primes = {2, 3, 5, 7};
+        AtomicLong remoteState = new AtomicLong(1);
+
+        Cache<Integer, Long> cache = EvictableCache.buildWith(CacheBuilder.newBuilder());
+        Integer key = 42;
+        int threads = 4;
+
+        CyclicBarrier barrier = new CyclicBarrier(threads);
+        ExecutorService executor = newFixedThreadPool(threads);
+        try {
+            List<Future<Void>> futures = IntStream.range(0, threads)
+                    .mapToObj(threadNumber -> executor.submit(() -> {
+                        // prime the cache
+                        assertEquals((long) cache.get(key, remoteState::get), 1L);
+                        int prime = primes[threadNumber];
+
+                        barrier.await(10, SECONDS);
+
+                        // modify underlying state
+                        remoteState.updateAndGet(current -> current * prime);
+
+                        // invalidate
+                        switch (invalidation) {
+                            case INVALIDATE_KEY:
+                                cache.invalidate(key);
+                                break;
+                            case INVALIDATE_PREDEFINED_KEYS:
+                                cache.invalidateAll(List.of(key));
+                                break;
+                            case INVALIDATE_SELECTED_KEYS:
+                                Set<Integer> keys = ((EvictableCache<Integer, ?>) cache).keySet().stream()
+                                        .filter(foundKey -> (int) foundKey == key)
+                                        .collect(toImmutableSet());
+                                cache.invalidateAll(keys);
+                                break;
+                            case INVALIDATE_ALL:
+                                cache.invalidateAll();
+                                break;
+                        }
+
+                        // read through cache
+                        long current = cache.get(key, remoteState::get);
+                        if (current % prime != 0) {
+                            fail(format("The value read through cache (%s) in thread (%s) is not divisable by (%s)", current, threadNumber, prime));
+                        }
+
+                        return (Void) null;
+                    }))
+                    .collect(toImmutableList());
+
+            futures.forEach(MoreFutures::getFutureValue);
+
+            assertEquals(remoteState.get(), 2 * 3 * 5 * 7);
+            assertEquals((long) cache.get(key, remoteState::get), remoteState.get());
+        }
+        finally {
+            executor.shutdownNow();
```

Comment:
I would keep executor as a field, that way each test would be less code would be less polluted.

Reply:
since threads are paramout here, i'd prefer the tests not to share the executor